// next-session/dist/compat.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.promisifyStore = exports.expressSession = void 0;
const util_1 = require("util");
const events_1 = require("events");
const memory_1 = __importDefault(require("./store/memory"));
function CompatibleStore() {
    // @ts-ignore
    events_1.EventEmitter.call(this);
}
util_1.inherits(CompatibleStore, events_1.EventEmitter);
// no-op for compat
function expressSession(options) { }
exports.expressSession = expressSession;
expressSession.Store = CompatibleStore;
function CallbackMemoryStore() { }
util_1.inherits(CallbackMemoryStore, CompatibleStore);
CallbackMemoryStore.prototype.get = util_1.callbackify(memory_1.default.prototype.get);
CallbackMemoryStore.prototype.set = util_1.callbackify(memory_1.default.prototype.set);
CallbackMemoryStore.prototype.destroy = util_1.callbackify(memory_1.default.prototype.destroy);
CallbackMemoryStore.prototype.all = util_1.callbackify(memory_1.default.prototype.all);
expressSession.MemoryStore = CallbackMemoryStore;
function promisifyStore(store) {
    console.warn('promisifyStore has been deprecated: express-session store still works without using this.');
    return store;
}
exports.promisifyStore = promisifyStore;
